### PR TITLE
Various refactors to the incr comp workproduct handling

### PR DESCRIPTION
**Pull request by bjorn3**
This is the result of me looking into adding support for having multiple object files for a single codegen unit to incr comp. This is necessary to support inline assembly in cg_clif without requiring partial linking which is not supported on Windows and seems to fail on macOS for some reason. Cg_clif uses an external assembler to handle inline asm and thus produces one object file with regular functions and one object file containing compiled inline asm for each codegen unit which uses inline asm. Current incr comp can't handle this. This PR doesn't yet add support for this, but it makes it easier to do so.

### CODE REVIEW

**Comment by nagisa** [comment]
@bors r+

Although it would have been nice if some of the commits (e.g. 81f73254bc5d5ae7436cb3d84664ccbd963654fe) specified why this change makes sense (“this function only ever returns work products from a previous compilation session” or something) in the description.

File: compiler/rustc_codegen_cranelift/src/driver/aot.rs [hunk 1]
```diff
@@ -84,21 +80,17 @@ fn reuse_workproduct_for_cgu(
...
-            tcx.sess.err(&format!(
-                "unable to copy {} to {}: {}",
-                source_file.display(),
-                obj_out.display(),
-                err
-            ));
-        }
+    let work_product = cgu.previous_work_product(tcx);
+    let obj_out = tcx.output_filenames(()).temp_path(OutputType::Object, Some(cgu.name().as_str()));
+    let object = Some(obj_out.clone());
```
Comment (klensy) [inline compiler/rustc_codegen_cranelift/src/driver/aot.rs:85]: Is it necessary to clone? Looks like `obj_out` can be directly passed to `CompiledModule`.

File: compiler/rustc_codegen_ssa/src/back/write.rs [hunk 2]
```diff
@@ -854,26 +854,23 @@ fn execute_copy_from_cache_work_item<B: ExtraBackendMethods>(
...
+        let diag_handler = cgcx.create_diag_handler();
+        diag_handler.err(&format!(
+            "unable to copy {} to {}: {}",
+            source_file.display(),
+            obj_out.display(),
+            err
+        ));
     }
 
     assert_eq!(object.is_some(), module_config.emit_obj != EmitObj::None);
```
Comment (klensy) [inline compiler/rustc_codegen_ssa/src/back/write.rs:876]: Same here, plus that assert.

File: compiler/rustc_codegen_cranelift/src/driver/aot.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_codegen_cranelift/src/driver/aot.rs:85]: RIght, will do that in a follow up.

File: compiler/rustc_codegen_ssa/src/back/write.rs (hunk 2 above)
Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/write.rs:876]: Same

**Comment by nagisa** [comment]
Perfect opportunity to address the comments above. r=me after.

**Comment by bjorn3** [comment]
Rebased, added rationale to commit messages and applied @klensy's suggestions.

@bors r=nagisa

**Pull request merged**